File: JavaFactorial/src/de/luschny/math/factorial/FactorialProductRecursive.java
// Copyright (C) 2004-2009 Peter Luschny, MIT License applies.
// See http://en.wikipedia.org/wiki/MIT_License
// Visit http://www.luschny.de/math/factorial/FastFactorialFunctions.htm
// Comments mail to: peter(at)luschny.de
package de.luschny.math.factorial;

import de.luschny.math.arithmetic.Xint;

public class FactorialProductRecursive implements IFactorialFunction {

    public FactorialProductRecursive() {
    }

    @Override
    public String getName() {
        return "ProductRecursive  ";
    }

    @Override
    public Xint factorial(int n) {
        if (n < 0) {
            throw new ArithmeticException("Factorial: n has to be >= 0, but was " + n);
        }

        if (1 < n) {
            return recProduct(1, n);
        }

        return Xint.ONE;
    }

    private Xint recProduct(int n, int len) {
        if (1 < len) {
            int l = len / 2;
            return recProduct(n, l).multiply(recProduct(n + l, len - l));
        }

        return Xint.valueOf(n);
    }
} // endOfFactorialProductRecursive

File: SilverFactorial64/Sharith/Factorial/FactorialProductRecursive.cs
// -------- ToujoursEnBeta
// Author & Copyright : Peter Luschny
// License: LGPL version 3.0 or (at your option)
// Creative Commons Attribution-ShareAlike 3.0
// Comments mail to: peter(at)luschny.de
// Created: 2010-03-01

namespace Sharith.Factorial 
{
    using XInt = Arithmetic.XInt;

    public class ProductRecursive : IFactorialFunction 
    {
        public string Name => "ProductRecursive    ";

        public XInt Factorial(int n)
        {
            if (n < 0)
            {
                throw new System.ArgumentOutOfRangeException(
                    this.Name + ": " + nameof(n) + " >= 0 required, but was " + n);
            }

            if (1 < n)
            {
                return this.RecProduct(1, n);
            }

            return XInt.One;
        }

        private XInt RecProduct(int n, int len)
        {
            if (1 < len)
            {
                var l = len >> 1;
                return this.RecProduct(n, l) * this.RecProduct(n + l, len - l);
            }

            return new XInt(n);
        }
    }
} // endOfFactorialProductRecursive
